server/src/main/java/io/crate/planner/optimizer/iterative/Memo.java

```diff
@@ -0,0 +1,225 @@
+
+        for (int group : references) {
+            Group childGroup = groups.get(group);
+            if (!childGroup.incomingReferences.remove(fromGroup)) {
+                throw new IllegalStateException("Reference to remove not found");
+            }
+            ;
+            if (childGroup.incomingReferences.isEmpty()) {
+                deleteGroup(group);
+            }
+        }
+    }
+
+    private Set<Integer> allReferences(LogicalPlan node) {
+        return node.sources().stream()
+            .map(GroupReference.class::cast)
+            .map(GroupReference::groupId)
+            .collect(Collectors.toSet());
+    }
+
+    private void deleteGroup(int group) {
+        LogicalPlan deletedNode = groups.remove(group).membership;
+        decrementReferenceCounts(deletedNode, group);
+    }
+
+    private LogicalPlan insertChildrenAndRewrite(LogicalPlan node) {
+        return node.replaceSources(
+            node.sources().stream()
+                .map(child -> new GroupReference(
+                    insertRecursive(child),
+                    child.outputs()))
+                .collect(Collectors.toList()));
+    }
+
+    private int insertRecursive(LogicalPlan node) {
+        if (node instanceof GroupReference) {
+            return ((GroupReference) node).groupId();
+        }
+
+        int group = nextGroupId();
+        LogicalPlan rewritten = insertChildrenAndRewrite(node);
+
+        groups.put(group, Group.withMember(rewritten));
+        incrementReferenceCounts(rewritten, group);
+
+        return group;
+    }
+
+    private int nextGroupId() {
+        return nextGroupId++;
+    }
+
+    public int groupCount() {
+        return groups.size();
+    }
+
+    private static final class Group {
+        static Group withMember(LogicalPlan member) {
+            return new Group(member);
+        }
+
+        private LogicalPlan membership;
+        private final List<Integer> incomingReferences = new ArrayList<>();
+
+        private Group(LogicalPlan member) {
+            this.membership = requireNonNull(member, "member is null");
+        }
+    }
+
+    public static LogicalPlan resolveGroupReferences(LogicalPlan node, GroupReferenceResolver lookup) {
+        requireNonNull(node, "node is null");
+        return node.accept(new ResolvingVisitor(lookup), null);
+    }
+
+    private static class ResolvingVisitor extends LogicalPlanVisitor<Void, LogicalPlan> {
+        private final GroupReferenceResolver lookup;
+
+        public ResolvingVisitor(GroupReferenceResolver lookup) {
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        public LogicalPlan visitPlan(LogicalPlan node, Void context) {
+            List<LogicalPlan> children = node.sources().stream()
+                .map(child -> child.accept(this, context))
+                .collect(Collectors.toList());
+
+            return node.replaceSources(children);
```

Comment:
The `replaceSources` implementations usually get the first element and rarely re-use the list as is so it might be cheaper to use `Lists2.mapLazy` here to only create 1 wrapper instead of copying the list (and further above as well)